src/back/glsl/mod.rs

```diff
@@ -2532,7 +2532,63 @@ impl<'a, W: Write> Writer<'a, W> {
                     Mf::Log2 => "log2",
                     Mf::Pow => "pow",
                     // geometry
-                    Mf::Dot => "dot",
+                    Mf::Dot => {
+                        let inner = ctx.info[arg].ty.inner_with(&self.module.types);
+                        match *inner {
+                            TypeInner::Vector { kind, size, .. } => {
+                                use crate::ScalarKind;
+                                match kind {
+                                    ScalarKind::Float => "dot",
+                                    ScalarKind::Sint | ScalarKind::Uint => {
+                                        // No intrinsic function for integer dot product in GLSL,
+                                        // transform the function call into an arithmetic expression
+                                        let arg1 =
+                                            arg1.ok_or_else(|| Error::Custom("Missing dot arg1".to_owned()))?;
+
+                                        write!(self.out, "(")?;
+
+                                        self.write_expr(arg, ctx)?;
```

Comment:
Writing the same expression multiple times is a valid thing to do, but it can end up with a lot of unnecessary repetition.
We need to force the arguments of `dot(integer, integer)` to be "baked".
Currently the baking logic in GLSL is just checking `bake_ref_count()`. Here, it needs to change to something like:
```rust
if ref_count > 1 || is_argument_of_integer_dot_product(expr) {
 // bake
}
```

Reply:
I tried to find a way to easily bake the arguments of a dot product, but to no avail. If I understand correctly, the "needs to be baked" information should live somewhere in the `FunctionCtx` that is created by each frontend (just like `named_expressions` are), so that when the time comes, the backend can know that the expression needs to be baked. This is a major change to Naga, so I don't think this is something I can achieve.

OTOH, I was wondering: would it be reasonable to do some "local" baking by creating a scope and using a temporary variable? I recon this is a lot of code that handles a very specific case, but perhaps it could be a good compromise.

Reply:
The way I thought this could work is you making a `BitSet` to store indices of expressions (within the currently processed functions) that are integer arguments to `dot()` product. When starting each function, you'd clear that bitset, then walk through all expressions and add the relevant indices to the bitset. And then `is_argument_of_integer_dot_product` is basically a check if `expr.index()` is in this bitset.

Reply:
I think I get it now, this would be a change that lives mostly only inside the `write_function` call of a backend. I will see if I can manage something reasonable!